File: FaceRecognition_With_EKF_Tracking/Implementation/movingFaceRecog.py
import cv2
import numpy as np
import dlib
from scipy.spatial import distance

# Initialize video capture (0 is the default camera)
cap = cv2.VideoCapture(0)

# Load dlib face detector and face recognition model
face_detector = dlib.get_frontal_face_detector()
face_recognizer = dlib.face_recognition_model_v1("dlib_face_recognition_resnet_model_v1.dat")
shape_predictor = dlib.shape_predictor("shape_predictor_68_face_landmarks.dat")

# Set up known face encoding (replace with a captured encoding for your target face)
known_face_encoding = None
def get_face_encoding(frame, detector, predictor, recognizer):
    # Detect faces in the color frame
    faces = detector(frame)
    if faces:
        shape = predictor(frame, faces[0])
        face_encoding = np.array(recognizer.compute_face_descriptor(frame, shape))
        return face_encoding
    return None

# If known_face_encoding is not set, capture a sample for it
while known_face_encoding is None:
    ret, frame = cap.read()
    if ret:
        known_face_encoding = get_face_encoding(frame, face_detector, shape_predictor, face_recognizer)
        cv2.imshow("Capture Known Face", frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

cap.release()
cv2.destroyAllWindows()

# EKF parameters
state = np.array([0, 0, 0], dtype=float)  # [x, y, theta]
P = np.eye(3) * 500  # Covariance matrix
R = np.diag([0.5, 0.5, np.deg2rad(5)])  # Motion noise covariance
Q = np.eye(2) * 2  # Measurement noise covariance

def predict(state, P, u):
    """Predict step of EKF."""
    F = np.eye(3)  # State transition model
    state = state + u  # Apply motion (u is [dx, dy, dtheta])
    P = F @ P @ F.T + R  # Update covariance
    return state, P

def update(state, P, z, H):
    """Update step of EKF."""
    predicted_observation = H @ state
    y = z - predicted_observation[:2]
    S = H @ P @ H.T + Q
    K = P @ H.T @ np.linalg.inv(S)
    state = state + (K @ y)
    P = (np.eye(3) - K @ H) @ P
    return state, P

# Initialize video capture again
cap = cv2.VideoCapture(0)

while cap.isOpened():
    ret, frame = cap.read()
    if not ret:
        break

    # Detect faces in the color frame
    faces = face_detector(frame)
    if faces:
        for face in faces:
            # Draw rectangle around face
            x, y, w, h = (face.left(), face.top(), face.width(), face.height())
            cv2.rectangle(frame, (x, y), (x+w, y+h), (255, 0, 0), 2)

            # Get encoding of the detected face
            shape = shape_predictor(frame, face)
            current_face_encoding = np.array(face_recognizer.compute_face_descriptor(frame, shape))

            # Compare with known face encoding
            if known_face_encoding is not None:
                match_distance = distance.euclidean(known_face_encoding, current_face_encoding)
                if match_distance < 0.6:  # Threshold for a match
                    cv2.putText(frame, "Trump", (x, y - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 1)
                    face_center = np.array([x + w / 2, y + h / 2])

                    # EKF Predict Step
                    u = np.array([0, 0, 0])  # Assuming no control input
                    state, P = predict(state, P, u)

                    # EKF Update Step (Use the face center as the measurement)
                    z = face_center  # Measurement
                    H = np.eye(2, 3)  # Measurement model
                    state, P = update(state, P, z, H)

                    # Display the estimated state on the frame
                    cv2.putText(frame, f"State: x={state[0]:.2f}, y={state[1]:.2f}, theta={np.degrees(state[2]):.2f}",
                                (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 255), 1, cv2.LINE_AA)
                else:
                    cv2.putText(frame, "Unknown Face", (x, y - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 1)

    # Show the current frame
    cv2.imshow('EKF-SLAM with Face Recognition', frame)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

cap.release()
cv2.destroyAllWindows()

File: FaceRecognition_With_EKF_Tracking/Implementation/movingFaceRecog.cpp
//
// Copyright(c) 2024 deepwave-ai. All Rights Reserved.
//
// faceEKFTracking.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <opencv2/opencv.hpp>
#include <dlib/opencv.h>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing.h>
#include <dlib/dnn.h>
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>

// Alias for dlib's face recognition model

using namespace dlib;
using namespace std;


template <template <int, template<typename> class, int, typename> class block, int N, template<typename> class BN, typename SUBNET>
using residual = dlib::add_prev1<block<N, BN, 1, dlib::tag1<SUBNET>>>;

template <template <int, template<typename> class, int, typename> class block, int N, template<typename> class BN, typename SUBNET>
using residual_down = dlib::add_prev2<dlib::avg_pool<2, 2, 2, 2, dlib::skip1<dlib::tag2<block<N, BN, 2, dlib::tag1<SUBNET>>>>>>;

template <int N, template <typename> class BN, int stride, typename SUBNET>
using block = BN<dlib::con<N, 3, 3, 1, 1, dlib::relu<BN<dlib::con<N, 3, 3, stride, stride, SUBNET>>>>>;

template <int N, typename SUBNET> using ares = dlib::relu<residual<block, N, dlib::affine, SUBNET>>;
template <int N, typename SUBNET> using ares_down = dlib::relu<residual_down<block, N, dlib::affine, SUBNET>>;

template <typename SUBNET> using alevel0 = ares_down<256, SUBNET>;
template <typename SUBNET> using alevel1 = ares<256, ares<256, ares_down<256, SUBNET>>>;
template <typename SUBNET> using alevel2 = ares<128, ares<128, ares_down<128, SUBNET>>>;
template <typename SUBNET> using alevel3 = ares<64, ares<64, ares<64, ares_down<64, SUBNET>>>>;
template <typename SUBNET> using alevel4 = ares<32, ares<32, ares<32, SUBNET>>>;

using anet_type = dlib::loss_metric<dlib::fc_no_bias<128, dlib::avg_pool_everything<
    alevel0<
    alevel1<
    alevel2<
    alevel3<
    alevel4<
    dlib::max_pool<3, 3, 2, 2, dlib::relu<dlib::affine<dlib::con<32, 7, 7, 2, 2,
    dlib::input_rgb_image_sized<150> >>>>>>>>>>>>;


// EKF Predict Step
void predict(cv::Mat& state, cv::Mat& P, const cv::Mat& u, const cv::Mat& R) {
    state += u;
    P = P + R;
}

// EKF Update Step
void update(cv::Mat& state, cv::Mat& P, const cv::Mat& z, const cv::Mat& H, const cv::Mat& Q) {
    cv::Mat predicted_observation = H * state;
    cv::Mat y = z - predicted_observation(cv::Range(0, 2), cv::Range::all());
    cv::Mat S = H * P * H.t() + Q;
    cv::Mat K = P * H.t() * S.inv();
    state += K * y;
    cv::Mat I = cv::Mat::eye(3, 3, CV_64F);
    P = (I - K * H) * P;
}

int main() {
    // Initialize video capture
    cv::VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error: Unable to open the camera." << std::endl;
        return -1;
    }

    // Load Dlib models
    dlib::frontal_face_detector face_detector = dlib::get_frontal_face_detector();
    dlib::shape_predictor shape_predictor;
    dlib::deserialize("shape_predictor_68_face_landmarks.dat") >> shape_predictor;
    anet_type face_recognizer;
    dlib::deserialize("dlib_face_recognition_resnet_model_v1.dat") >> face_recognizer;

    // Capture a reference face encoding
    dlib::matrix<float, 0, 1> known_face_encoding;
    bool reference_face_captured = false;
    while (!reference_face_captured) {
        cv::Mat frame;
        cap >> frame;
        if (frame.empty()) break;

        dlib::cv_image<dlib::bgr_pixel> dlib_frame(frame);
        std::vector<dlib::rectangle> faces = face_detector(dlib_frame);

        if (!faces.empty()) {
            dlib::full_object_detection shape = shape_predictor(dlib_frame, faces[0]);
            // Align face and create a face chip
            matrix<rgb_pixel> face_chip;
            extract_image_chip(dlib_frame, get_face_chip_details(shape, 150, 0.25), face_chip);

            known_face_encoding = face_recognizer(face_chip);
            reference_face_captured = true;
        }

        cv::imshow("Capture Reference Face", frame);
        if (cv::waitKey(1) == 'q') break;
    }

    // EKF initialization
    cv::Mat state = cv::Mat::zeros(3, 1, CV_64F); // [x, y, theta]
    cv::Mat P = cv::Mat::eye(3, 3, CV_64F) * 500;

    cv::Mat R = (cv::Mat_<double>(3, 3) << 0.5, 0, 0,
        0, 0.5, 0,
        0, 0, 5.0 * CV_PI / 180.0);

    cv::Mat Q = cv::Mat::eye(2, 2, CV_64F) * 2;

    while (true) {
        cv::Mat frame;
        cap >> frame;
        if (frame.empty()) break;

        dlib::cv_image<dlib::bgr_pixel> dlib_frame(frame);
        std::vector<dlib::rectangle> faces = face_detector(dlib_frame);

        for (const auto& face : faces) {
            int x = face.left();
            int y = face.top();
            int w = face.width();
            int h = face.height();
            cv::rectangle(frame, cv::Rect(x, y, w, h), cv::Scalar(255, 0, 0), 2);

            dlib::full_object_detection shape = shape_predictor(dlib_frame, face);
            matrix<rgb_pixel> face_chip;
            extract_image_chip(dlib_frame, get_face_chip_details(shape, 150, 0.25), face_chip);

            dlib::matrix<float, 0, 1> current_face_encoding = face_recognizer(face_chip);

            double match_distance = dlib::length(current_face_encoding - known_face_encoding);
            if (match_distance < 0.6) {
                cv::putText(frame, "Matched", cv::Point(x, y - 10), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 0), 1);

                cv::Mat face_center = (cv::Mat_<double>(2, 1) << x + w / 2.0, y + h / 2.0);

                // EKF Predict and Update
                cv::Mat u = cv::Mat::zeros(3, 1, CV_64F); // No control input
                predict(state, P, u, R);
                cv::Mat H = cv::Mat::eye(2, 3, CV_64F);
                update(state, P, face_center, H, Q);

                std::string state_text = "State: x=" + std::to_string(state.at<double>(0)) + ", y=" +
                    std::to_string(state.at<double>(1)) + ", theta=" +
                    std::to_string(state.at<double>(2) * 180.0 / CV_PI);

                cv::putText(frame, state_text, cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 255, 255), 1);
            }
            else {
                cv::putText(frame, "Unknown", cv::Point(x, y - 10), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 255), 1);
            }
        }

        cv::imshow("EKF-SLAM with Face Recognition", frame);
        if (cv::waitKey(1) == 'q') break;
    }

    cap.release();
    cv::destroyAllWindows();

    return 0;
}
